Metadata workflow

pcm32 opened the pull request:
This PR reorganises the setup to have a data and a metadata Snakemake workflows

- Adds the metadata workflow based on https://github.com/ebi-gene-expression-group/MAGE-Tab-merger repository, 
- Utilises generated dataset in data analysis workflow

pcm32 committed:
Reorganisation and working metadata wf

pcm32 committed:
Almost working

pcm32 committed:
Fix column usage for transcripts.

pcm32 committed:
Documentation and clean up.

pcm32 committed:
Use txt for SDRF.

pcm32 committed:
Add GTF consumption

pcm32 committed:
Add gitignore

pcm32 committed:
Refgenie env

pcm32 committed:
Use config instead of env vars and introduce targets

Commit:
Fixes Snamefile bugs stopping it from execution

anilthanki committed:
removes config file added by mistake in last commit

anilthanki committed:
replaces soft link with hard link, soft link was failing run

anilthanki committed:
ads stringent filter for grepping species and entries with gtf, also removes comma(,) from alias if present

anilthanki committed:
adds analysis-methods rule, rearrange output dir, unzips GTF from RefGenie for next steps

anilthanki committed:
adds cache_path config for input and output files, adds batch config

anilthanki committed:
updates version of r-atlas-internal and adds perl-xml-parser

anilthanki committed:
updates code to compatible with S4 class, drops unused levels for batch factor

anilthanki committed:
removes trap command and redirects logs to log file

anilthanki committed:
adds updated parseSDRF function to add Study factor in parsed SDRF

anilthanki committed:
Update README.md

anilthanki committed:
Update README.md

irisdianauy commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,106 @@
+    threads: 2
+    shell:
+        """
+	irap_gtf2featlength --gtf {input.gtf} --out {params.lengths} --cores {threads}
```
Has running this part finally worked?  Have you also tried using the singularity container for irap?

irisdianauy replied:
[This is how ISL calls the iRAP container](https://github.com/ebi-gene-expression-group/isl/blob/develop_codon/lib/irap.sh), and relevant variables are set in the ISL env.

anilthanki replied:
Yes I managed to get it working with conda env

pmb59 commented on README.md:
````diff
@@ -1,35 +1,136 @@
+To run the metadata part you will need to specify the following environment variables: **Not using Environment variables in new setup, but using Snakemake config instead.**
+
+```
+# The accessions of experiments that you want to merge
+export ACCESSIONS="E-MTAB-4395,E-MTAB-4342,E-MTAB-4128"
````
I assume all accessions need to come from same the species. Could you implement a check in the Snakefile that confirms this? 

A related question. Do all accessions here are 'baseline' RNA-seq in Atlas, right? Could this also be checked so that microarrays or proteomics experiments are not included by mistake? This info is available in the XML, an option could be using something like [this](https://github.com/ebi-gene-expression-group/experiment_metadata/blob/develop/zooma-mappings-wf/rules/common.smk#L42)

anilthanki replied:
Thanks.. yes all accession needs to be from same species, 

I will add checks for species and experiment type

anilthanki replied:
Added fix in [d66c44e](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/d66c44e590f3bcc7f68fe71ff0767cb144aa0e9c)

pmb59 commented on README.md:
```diff
@@ -1,35 +1,136 @@
+E-SUPR-1-configuration.xml                      E-SUPR-1.condensed.sdrf.tsv
+E-SUPR-1-raw-counts.tsv.undecorated             E-SUPR-1.sdrf.tsv
+E-SUPR-1-transcripts-raw-counts.tsv.undecorated	E-SUPR-1.selected_studies.txt
+E-SUPR-1.gtf                                    E-SUPR-1-analysis-methods.tsv
```
what is stored in the GTF file?  is it generated by the workflow? If it is just copied from refgenie and not altered in any way, I am wondering if we should keep the original filename.

anilthanki replied:
It's copied from RefGenie, we can use original file name its just that then it needs to supplied to second workflow as a parameter..

pmb59 replied:
can you add the original GTF filename to the Methods file? (if not done yet)

pmb59 commented on README.md:
```diff
@@ -1,35 +1,136 @@
-The current analysis setup enables to merge Atlas RNA-Seq baseline datasets that have the required compatibility in terms of metadata. To merge the metadata of datasets, you should use the [MAGE-Tab-merger](https://pypi.org/project/MAGE-Tab-merger/) package to produce:
+## Metadata
+
+To merge the metadata of datasets, we use within a Snakemake workflow the [MAGE-Tab-merger](https://pypi.org/project/MAGE-Tab-merger/) package to produce:
```
Is it possible to merge IDFs? the MAGE-Tab Merger package currently says 

> Note: IDF merging is still work in progress.

Do we need merge a merged IDF?

anilthanki replied:
I dont see much progress on IDF merging after that comment...

pmb59 commented on README.md:
```diff
@@ -1,35 +1,136 @@
 # Atlas baseline meta-analysis
 
```
```suggestion
The repository contains workflows for metadata generation and data analysis.
```

pmb59 commented on README.md:
```diff
@@ -1,35 +1,136 @@
+export CACHE_PATH="/absolute/path/to/cache"
+# The name used to the batch category, usually "study"
+export BATCH=study
+# The covariate used to merge metadata, as written in the SDRF/Condensed
+export COVARIATE='organism part'
```
a bit of explanation below about this parameter could help. What are the consequences of using different COVARIATEs?

anilthanki replied:
Covariates are the factors which may affect the final results regardless of process followed. 

Here we are using `organism part`, one another example of covariate could be biological condition

pmb59 commented on data-analysis/bin/get_assay_groups.pl:
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env perl
```
Maybe you can add [atlas-bash-util](https://github.com/ebi-gene-expression-group/atlas-bash-util) as a submodule and use `get_assay_groups.pl` from there?
(if you do, please double-check that the code is actually the same )

Same comment for some other scripts that are here https://github.com/ebi-gene-expression-group/atlas-analysis  (if appropriate)

anilthanki replied:
Looks like it was added in older version of pipeline but not used.. removed along with some other un-used scripts in [03cb628](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/03cb628ebcec7186e6ad865442fd54f883d301cd)

pmb59 commented on data-analysis/bin/quantile_normalize.sh:
```diff
@@ -35,11 +35,13 @@ done
 tmp_files="$path_to_destination.$(date -u +"%Y-%m-%d_%H:%M").tmp"
-trap 'rm -f ${tmp_files}.*' INT TERM EXIT
+#trap 'rm -f ${tmp_files}.*' INT TERM EXIT
 
```
Could you rename this script `quantile_normalize_meta.sh` (or similar) so we don't confuse it with the one here? https://github.com/ebi-gene-expression-group/atlas-analysis/tree/develop/norm

anilthanki replied:
fixes in [d5a1a00](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/d5a1a00fb461a24d974bd526ed88518a004eab7b)

pmb59 commented on data-analysis/bin/quantile_normalize.sh:
```diff
@@ -35,11 +35,13 @@ done
 
 tmp_files="$path_to_destination.$(date -u +"%Y-%m-%d_%H:%M").tmp"
-trap 'rm -f ${tmp_files}.*' INT TERM EXIT
+#trap 'rm -f ${tmp_files}.*' INT TERM EXIT
```
why this is not necessary anymore?

anilthanki replied:
Pipeline was failing with this command in place.

anilthanki replied:
Added trap command back in [66803b3](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/66803b30f9b2d9726c7c1f373319fdee7dc0ba63)

pmb59 commented on meta-data/Snakefile:
```diff
@@ -0,0 +1,236 @@
+        retrieve_data.py -i {output.data_path} -a {params.accessions} -f {params.optionals} -d 2> {log}
+        """
+
+def existing_or_retrieve(type):
```
this could be moved above with other helper functions

anilthanki replied:
Fixes in [71f1946](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/71f1946fdec7dfd021a41070a2a92af520a4ba58)

pmb59 commented on meta-data/Snakefile:
```diff
@@ -0,0 +1,236 @@
+        if type == "path":
+            return rules.retrieve_data.output.data_path
+        else:
+            return rules.retrieve_data.output.data
```
I think you should do here:
```
 if type == "path":
else if type == "data":
else
```
to handle explicitly the type "data"

anilthanki replied:
Fixes in [555bd20](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/555bd201cac7167c6113b2859f1620589aa0adb8)

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,106 @@
+    shell:
+        "cp {input.raw_counts_file} {output.labeled_genes_raw_counts}"
+
+rule lengths:
```
could you rename this rule so the name is more informative?

anilthanki replied:
Fixes in [6104229](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/6104229d0e09f756e0bdb3e999f52a1ec4a49c27)

anilthanki replied:
adds renamed file in [7925cdb](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/7925cdb7a688d2b6d6df6be16ea3e141062aac1e)

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,106 @@
+ANNOT_OBJECTS=["gene", "transcript"]
+METRICS=['tpm', 'fpkm']
```
Not all accessions have `tpm` or `fpkm` metrics. In some (rare cases) it is only one of them. Optionally, you could implement a check for that

anilthanki replied:
here `irap_raw2metric ` rule generates `tpm` and `fpkm` matrices from counts for the accessions

pmb59 commented on meta-data/Snakefile:
```diff
@@ -0,0 +1,236 @@
+        new_accession=config["new_accession"] 
+    shell:
+        """
+        echo "Pipeline version\t<a href=\"https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/\">Atlas baseline meta-analysis</a>" > {output}
+        echo "Input Accessions\t" {params.accessions} >> {output}
+        echo "Output accession\t" {params.new_accession} >> {output}
```
In bulk analysis, the file -analysis-methods.tsv has the actual analysis methods for each accession.
Here we don't have that. I wonder if this file should be named -analysis-methods.tsv or something different.

Maybe you could add also here species, and covariate information used for the meta-analysis

anilthanki replied:
In discussion with Pablo, he mentioned that we should generate our own analysis-methods file for this describing the process we followed, 

I am planning to add bit description of steps the process follows, I will try to replicated bulk analysis analysis-methods.tsv

anilthanki replied:
added some more info in [3b5f7f8](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/3b5f7f82bbdc760f88e098a3ab72e00892076294)

pmb59 replied:
You could add here to the versions of IRAP and MAGETab-merger.

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,106 @@
+            --tsv_corrected_counts {output.tsv_corrected_counts}
+        """
+
+rule irap_raw2metric:
```
```suggestion
rule irap_raw2metric:
    """
    Generates TPM and FPKM metrics from counts for accessions 
    """
```

pmb59 commented on data-analysis/bin/batch_correction_v2.R:
```diff
@@ -62,6 +62,9 @@ correct_batch_effect<-function(experiment, model, method=c('ComBat','RUV','MNN')
+experimentSummary$batch <- droplevels(experimentSummary$batch)
+
+
 batch_corrected<-correct_batch_effect(experiment = experimentSummary, model= ~organism_part, method='ComBat')
```
why "Combat" is used and not others?

anilthanki replied:
`Combat` was chosen as a batch_correction method by predecessors but looking at [this paper](https://www.biorxiv.org/content/10.1101/2020.05.22.111161v2.full), looks like it was a good choice. 

`MNN` and `RUV` are provided as alternatives

pmb59 replied:
you could also add this info to the methods

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,106 @@
+            -d {output.quantile_result}
+        """
+
+rule transform2R:
```
I would recommend to add log files to the rules in this Snakefile

anilthanki replied:
adds logs in data-analysis workflow in commit [5efeb2a](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/5efeb2af09fb7f12d2ca8086a48b6b840e18f0e1)

pmb59 reviewed (changes requested):
It looks very good @anilthanki 

In addition to the comments I have left, memory allocation based on retrials could be implemented in case merging /analysing dozens/hundreds of datasets requires high RAM mem.

Future work you could address, perhaps in other PRs:
1 - adding CI (via github actions) that run snakemake with test data. E.g. merging two accessions
2- monitor mage-tab-merger version > 0.0.4. If IDF merging is provided, it would be good to have have it too, as the formal definition MAGE-TAB requires IDF (Investigation Description Format) and SDRF (Sample and Data Relationship Format).

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,106 @@
+ANNOT_OBJECTS=["gene", "transcript"]
+METRICS=['tpm', 'fpkm']
+
```
I usually include this in my workflows:
```suggestion
# set minimum snakemake version
min_version("6.6.1")

```

anilthanki committed:
Update README.md

Co-authored-by: Pedro Madrigal <8195212+pmb59@users.noreply.github.com>

anilthanki committed:
Defines minimum Snakemake version required

Co-authored-by: Pedro Madrigal <8195212+pmb59@users.noreply.github.com>

anilthanki committed:
adds filter based on experiment type and species

anilthanki committed:
rearranges helper function

anilthanki committed:
renames Snakemake rule

anilthanki committed:
renames quantile_normalize

anilthanki committed:
Explicitly sets data type

anilthanki committed:
adds log files

anilthanki committed:
adds renamed files

anilthanki committed:
removes unnecessary scripts

anilthanki committed:
adds more info in analysis-methods

anilthanki committed:
avoids RefGenie warning when limiting grep to 1

anilthanki committed:
Update data-analysis/Snakefile

Co-authored-by: Pedro Madrigal <8195212+pmb59@users.noreply.github.com>

anilthanki committed:
adding trap command back

anilthanki committed:
Merge branch 'feature/metadata-workflow' of https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis into feature/metadata-workflow

pmb59 committed:
add curation folder

anilthanki committed:
adds covariate param to be used in future

anilthanki committed:
Merge branch 'feature/metadata-workflow' of https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis into feature/metadata-workflow

anilthanki committed:
re-organises output adds output config variable, compatible with Jenkins

anilthanki committed:
refactors variable names

anilthanki committed:
changes output variable and redirects log files

anilthanki committed:
adds known bug and Jenkins integration

anilthanki committed:
removes print statements

anilthanki committed:
adds configurable covariate

anilthanki committed:
retrieves GTF using species name

anilthanki committed:
removes RefGenie bug info

anilthanki committed:
Removing ENV VARs information

Variables are set from Snakemake configuration, not using environment variables anymore

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,120 @@
+ANNOT_OBJECTS=["gene", "transcript"]
+METRICS=['tpm', 'fpkm']
+# set minimum snakemake version
+from snakemake.utils import min_version
+min_version("6.6.1")
```
I would move this chunk to the top of the script

anilthanki replied:
Fixed in [23ae1c3](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/23ae1c39f188c4bf284bf23741e49a11b36469ea)

pmb59 commented on README.md:
````diff
@@ -1,38 +1,101 @@
+E-SUPR-1.gtf                                    E-SUPR-1-analysis-methods.tsv
+```
+
+Most files are self explanatory, the `selected_studies.txt` files contains a comma separate list of the original accessions used.
````
```suggestion
Most files are self explanatory, the `selected_studies.txt` files contains a comma separate list of the original accessions used. The `gtf` is copied from Refgenie references.
```

pmb59 commented on README.md:
```diff
@@ -1,38 +1,101 @@
+
+### Jenkins job
+
+Jenkins job for this part of workflow is at http://gene-expression.ebi.ac.uk/jenkins/job/bulk_baseline_meta_analysis_dataprep/ 
```
Although unauthorised users won't have access to this link, we usually don't give jenkins information across public Atlas repos.  I would suggest to write an internal SOP for baseline meta-analysis where this info can be specified and detailed.

The SOP could also indicate the role of curators in this processes.

anilthanki replied:
Jenkins link removed

pmb59 commented on curation/README.md:
```diff
@@ -0,0 +1 @@
+# Selection of studies that can be merged in mete-analysis
```
```suggestion
# Selection of studies that can be merged in meta-analysis
```

pmb59 replied:
what's the purpose of this file? does it need editing by curators?

pmb59 commented on meta-data/Snakefile:
```diff
@@ -0,0 +1,279 @@
+    if "covariate_skip_values" in config:
+        optionals=f" --covariate-skip-values '{config['covariate_skip_values']}'"
+
+    return optionals
```
is it fine for downstream processes to return `optionals=""`?

anilthanki replied:
Its returning empty string so I don't see any issue

pmb59 commented on meta-data/Snakefile:
```diff
@@ -0,0 +1,279 @@
+    and retrieve_data config variables.
+    """
+    optionals=""
+    if "replace_downloads" in config:
+        optionals=" -r"
+
+    if "retrieve_data" in config or ('target' in config and config['target'] == TARGET_ANALYSIS_POST_CURATION):
```
what if both options are in `config`?

anilthanki replied:
It should consider `retrieve_data` and continue

pmb59 commented on meta-data/Snakefile:
```diff
@@ -0,0 +1,279 @@
+	species=config["species"]
+    shell:
+        """
+        echo "Pipeline version\t<a href=\"https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/\">Atlas baseline meta-analysis</a>" > {output}
+        echo "Input Accessions\t" {params.accessions} >> {output}
+	echo "Accession used in Analysis\t" {params.filtered_accession} >> {output}
+        echo "Output accession\t" {params.new_accession} >> {output}
+	echo "Retrieved GTF using RefGenie for the species\t" {params.species} >> {output}
```
Maybe you could add here the: 
- MAGETab-merger version
- Refgenie version

anilthanki replied:
Versions not specified in conda env to be able to use latest available. We can either specify and use hardcoded version or leave it as it is..

As both are separate conda env cant use both in one rule and will need to detect and store version in global var

pmb59 commented on data-analysis/bin/transform2R.r:
```diff
@@ -0,0 +1,650 @@
+#!/usr/bin/env Rscript
+
```
this R script needs abundant style editing

anilthanki replied:
Fixed in [00e7c4d](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/00e7c4d80ec2b42cae2ab76d6f021f0039d7eb78)

pmb59 commented on data-analysis/Snakefile:
```diff
@@ -0,0 +1,120 @@
+            --output {output.rdata} > {log}
+        """
+
+rule batch_correction_v2:
```
why is  named`_v2`?

anilthanki replied:
It came from predecessors, its good idea to change it

anilthanki replied:
fixed in 477e0c5 and [6bf98cb](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/6bf98cbabeb95b290ae2fed7f1b2443f0491e6c4)

pmb59 commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "data-analysis/bin/atlas-bash-util"]
```
where is this submodule actually used in the repo?

anilthanki replied:
It was added to replace some scripts but they were never used so removed submodule as well, removed in [8469691](https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/pull/1/commits/84696915bf3702ffbaa170dfe38432270ee046a3)

pmb59 commented on README.md:
```diff
@@ -1,38 +1,101 @@
 - tmp_results: main results, with all calculations for the merged dataset.
 - lengths: where lengths per gene, exon and transcripts are produced, based on the reference GTF file provided. These are used to normalise to TPMs and FPKMs.
+
+### Jenkins job
+
+Jenkins job for this part of workflow is at http://gene-expression.ebi.ac.uk/jenkins/job/bulk_baseline_meta_analysis/
```
Same as before, better move this to SOP

anilthanki replied:
Jenkins link removed

pmb59 reviewed (commented):
@anilthanki LGTM, almost production-ready!

I've left few comments

anilthanki committed:
Update README.md

Co-authored-by: Pedro Madrigal <8195212+pmb59@users.noreply.github.com>

anilthanki committed:
Update curation/README.md

Co-authored-by: Pedro Madrigal <8195212+pmb59@users.noreply.github.com>

anilthanki committed:
removes Jenkins jobs info

anilthanki committed:
Rename batch_correction_v2.R to batch_correction.R

anilthanki committed:
Update Snakefile

anilthanki committed:
Update Snakefile

anilthanki committed:
lint transform2R.r

anilthanki committed:
removes submodule

anilthanki committed:
removes submodule

anilthanki committed:
fixes file name change err

pmb59 commented on meta-data/Snakefile:
```diff
@@ -69,6 +69,42 @@ def get_index_column(suffix):
+        exp_type = []
+        config_tag = xmldoc.getElementsByTagName("configuration")
+        for i in config_tag:
+            if "rnaseq" in i.getAttribute("experimentType"):
```
experimentType can be
- 'rnaseq_mrna_baseline'
- 'rnaseq_mrna_differential'

Maybe you want to check here explicitly that the experiments belong to the first case?

pmb59 reviewed (approved):
@anilthanki LGTM!

I've added few other comments.

To avoid overloading this excellent first PR I would recommend that, if your tests are working, merge the PR and list in a new Issue in the repo those potential new features that can be addressed in the future (known covariate bug, improve methods file, standardisation with respect to current bulk pipelines, details on how to select studies, etc.)

anilthanki commented:
Thanks @pmb59 .. I created https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/issues/2 and https://github.com/ebi-gene-expression-group/atlas-baseline-meta-analysis/issues/3 from comments.. 

Feel free to add more info on them
